src/org/sosy_lab/java_smt/api/UserPropagator.java

```diff
@@ -0,0 +1,114 @@
+   * {@link PropagatorBackend#notifyOnFinalCheck()}.
+   */
+  void onFinalCheck();
+
+  /**
+   * This callback is invoked if the solver gets to know the value of a registered expression
+   * ({@link #registerExpression}).
+   * Within the callback, the user can raise conflicts via
+   * {@link PropagatorBackend#propagateConflict}, propagate consequences via
+   * {@link PropagatorBackend#propagateConsequence}, or influence the solvers decision heuristics
+   * via {@link PropagatorBackend#propagateNextDecision}.
+   * <p>
+   * The reported value is only known on the current and later push levels,
+   * but will get invalidated when backtracking.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnKnownValue()}.
+   *
+   * @param expr  The expressions whose value is known.
+   * @param value The value of the expression.
+   */
+  void onKnownValue(BooleanFormula expr, boolean value);
+
+  /**
+   * This callback is invoked if the solver decides to branch on a registered expression.
+   * ({@link #registerExpression}).
+   * Within the callback, the user can change the decision by calling
+   * {@link PropagatorBackend#propagateNextDecision}.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnDecision()}.
+   * @param expr The expressions whose value gets decided (usually a literal).
+   * @param value The decision value.
+   */
+  void onDecision(BooleanFormula expr, boolean value);
+
+  /**
+   * Connects this user propagator with a {@link PropagatorBackend}. The backend is used
+   * to register expressions, raise conflicts, propagate consequences, etc.
+   * @param backend The propagator backend.
+   */
+  void injectBackend(PropagatorBackend backend);
+
+  /**
+   * This method is similar to a constructor but is guaranteed to get invoked only after
+   * {@link #injectBackend} was successfully called.
+   * The user can enable notifications by accessing the injected {@link PropagatorBackend}.
+   */
+  void initialize();
+
+  /**
+   * Registers an expression to be observed by the {@link UserPropagator}.
+   * Solver actions related to the expression are reported:
+   * <ul>
+   *   <li>
+   *     The callback {@link UserPropagator#onKnownValue} gets invoked if the value of a
+   *     registered expression becomes known (if notification was enabled via
+   *    {@link PropagatorBackend#notifyOnKnownValue}).
+   *   </li>
+   *   <li>
+   *     The callback {@link UserPropagator#onDecision} gets invoked right before the
+   *     solver decides on the value of a registered expression (if notification was enabled via
+   *     {@link PropagatorBackend#notifyOnDecision()}).
+   *   </li>
+   * </ul>
+   *
+   * @param theoryExpr The expression to observe.
+   */
+  void registerExpression(BooleanFormula theoryExpr);
```

Comment:
Does this expression need to be a symbol (variable) or can it be a complex boolean formula?
Can we also register non-boolean symbols?

Examples:
- Can we register to "a&&b"? Would the propagator then be called for each operation involving "a&&b" in that exact combination/operation?
- Can we register "counter==42"? Would the propagator then be called for each operation involving "counter"?

Reply:
You can register any boolean expression, even `counter==42`. However, the propagator will notify on the value of the boolean expression and not on any variables inside. Basically, the propagator works on the boolean skeleton of the formula.
So you cannot track the precise value of `counter` and only indirectly know about its value by tracking atoms/predicates over `counter.`

Also, the propagator is more on the semantic level rather the syntactic one. If your whole asserted formula is `a || b` but you registered `a && b`, the propagator will still notify you once its partially found model fixes the value of `a && b`.
In this case, the solver will likely do this:
```
Decide a = false
Unit propagate b = true
Report (a&&b) == false to UserPropagtor
```
You can run the `SimpleUserPropagator` example I added which showcases various interactions and prints the solver steps involved.